src/coffeescript.coffee

```diff
@@ -271,111 +246,141 @@ parser.yy.parseError = (message, {token}) ->
   # from the lexer.
   helpers.throwSyntaxError "unexpected #{errorText}", errorLoc
 
-# Based on http://v8.googlecode.com/svn/branches/bleeding_edge/src/messages.js
-# Modified to handle sourceMap
-formatSourcePosition = (frame, getSourceMapping) ->
-  filename = undefined
-  fileLocation = ''
-
-  if frame.isNative()
-    fileLocation = "native"
-  else
-    if frame.isEval()
-      filename = frame.getScriptNameOrSourceURL()
-      fileLocation = "#{frame.getEvalOrigin()}, " unless filename
+if process.execArgv.includes('--enable-source-maps') or process.env.NODE_OPTIONS?.includes('--enable-source-maps')
+  registerCompiled = ->
```

Comment:
Don’t we want to only not monkey-patch `Error.prepareStackTrace`, not remove all of `registerCompiled`, when `--enable-source-maps` is set? `registerCompiled` does other stuff beyond source maps, it looks like.

Reply:
I've only found it used in relation to source maps though the comment seems to indicate otherwise.

Reply:
The `sources` object seems to be used as a local cache of compiled output, so that CoffeeScript doesn’t recompile the same file. And there’s the comment “This is exported to enable an external module to implement caching of compilation results” which I’m not sure about, but seems to imply that this cache both needs to be present and exported for other libraries to use.

Reply:
I've used `CoffeeScript.registerCompiled` in my fork of CoffeeCoverage to cache source maps. I can't find anywhere in the source where it uses the cached files outside of source mapping.

I believe that previously there was some "cleverness" where it would try to cache files to avoid recompiling but I think that was removed and the comment wasn't updated.

Reply:
> I believe that previously there was some “cleverness” where it would try to cache files to avoid recompiling but I think that was removed and the comment wasn’t updated.

Well, or the caching happens in an external package, like a build tool or build tool plugin that imports the `coffeescript` package. Which I think is the point, since the cache is exported?

Or put another way, is there any reason not to scope this PR as narrowly as possible? As in, why shouldn’t we use the `--enable-source-maps` check to decide whether to patch `Error.prepareStackTrace`, and limit the effect of `--enable-source-maps` to only that?

Reply:
From what I could see `registerCompiled` is the only relevant export that I could find. The `sources` and `sourceMaps` caches are not exported.

My primary reason for not doing any caching with the feature disabled is that since it only seems to be used for the source mapping it would be an unnecessary and possibly large amount of resources being used for no visible effect. Some files are hundreds of kb and some projects have hundreds of files so it could add up. I'm happy to reverse if I've misunderstood or we find out some other use case where this has an impact.

One other thing that I noticed is that we may also want to always add the inline map when `--enable-source-maps` is present so that the native node source mapping will work. I'll need to do some more testing running it with different environment configurations to make sure that all the parts fit together properly. Ideally I'll even add a test that checks the proper line number 😎

Reply:
> From what I could see `registerCompiled` is the only relevant export that I could find. The `sources` and `sourceMaps` caches are not exported.

Okay, fair enough. And though this diff is huge, I assume you didn’t change anything about the “normal” `registerCompiled` other than indent it?

> One other thing that I noticed is that we may also want to always add the inline map when `--enable-source-maps` is present so that the native node source mapping will work. I’ll need to do some more testing running it with different environment configurations to make sure that all the parts fit together properly. Ideally I’ll even add a test that checks the proper line number 😎

Yes please! There have been various issues and PRs opened over the years regarding source maps line numbers. I feel like every time I fix one bug related to line numbers I create another.

Another thing to consider is that Node 12 reaches end-of-life at the end of this month, and `--enable-source-maps` was added in 12.12 / 14.18. So CoffeeScript could drop support for Node < 14, and if there’s anything in Node’s internal source maps support that we could use, we could use that instead of doing it ourselves. Like patching `Error.prepareStackTrace` might never be necessary, for example, because we could just update our docs to tell users to run `--enable-source-maps` for source maps in Node, including correct stack traces, rather than us adding all that. I don’t know if there’s anything else in https://nodejs.org/dist/latest/docs/api/module.html#source-map-v3-support that would let us remove any more of our source maps-related code, but it might be worth a look.